### ISSUE TITLE
NaN when calculating derivatives of torch.erf

### ISSUE THREAD

**Issue by maybeLee** [pytorch/pytorch#91567]
### 🐛 Describe the bug

I was trying some loss functions and find the third derivative of `torch.erf` is NaN, which I think may be incorrect. The equation I wrote is `torch.erf(torch.log(torch.exp(torch.reciprocal(y+x)+z))+x+y)` and the input for x, y, z is `100., 70.78463832, 268.13439677`. I found the third derivative of `x, y, z` are NaNs. However, if I simplify the equation to `torch.erf(torch.reciprocal(y+x)+z+x+y)`, which is mathematical equivalent to the previous one because `log(exp(x))=x`, the third derivative of `x, y, z` are 0.0. Here is the code to reproduce this issue:

```
import torch

x = torch.tensor(100., requires_grad=True)
y = torch.tensor(70.78463832, requires_grad=True)
z = torch.tensor(268.13439677, requires_grad=True)

res_1 = torch.erf(torch.log(torch.exp(torch.reciprocal(y+x)+z))+x+y)
res_2 = torch.erf(torch.reciprocal(y+x)+z+x+y)
# res_1 and res_2 are mathematical equivalent.

def get_third(res):
  first = torch.autograd.grad(res, (y,), create_graph=True)
  second = torch.autograd.grad(first, (y,), create_graph=True)
  third = torch.autograd.grad(second, (y,), create_graph=True)
  return third

print(f"Forward Result. res_1: {res_1}, res_2: {res_2}")
print(f"Third Order Derivative Of y. res_1: {get_third(res_1)}, res_2: {get_third(res_2)}")

------ outputs ------
Forward Result. res_1: 1.0, res_2: 1.0
Third Order Derivative Of y. res_1: (tensor(nan, grad_fn=<AddBackward0>),), res_2: (tensor(0., grad_fn=<AddBackward0>),)
```

For your reference, I also try the same equation on TensorFlow to calculate the third derivative and TensorFlow outputs 0.0.

```
import tensorflow as tf

x = tf.Variable(100.)
y = tf.Variable(70.78463832)
z = tf.Variable(268.13439677)

with tf.GradientTape() as g2:
  with tf.GradientTape() as g1:
    with tf.GradientTape() as g0:
      g0.watch(y)
      res_1 = tf.math.erf(tf.math.reciprocal(y+x)+z+x+y)
    first = g0.gradient(res_1, y)
  second = g1.gradient(first, y)
third = g2.gradient(second, y)
    
print(f"Third Order Derivative Of y. res_1: {third}")

------ outputs ------
Third Order Derivative Of y. res_1: 0.0
```

I also provide a gist to reproduce the result: https://colab.research.google.com/drive/1LlT9hqIeqDfl0koxJLx3_lbVM7bS1jPt?usp=sharing

### Versions

1.13.0+cu116

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @kshitij12345

**Comment by mingfeima** [comment]
@min-jean-cho , could you please check how TF fix this type of issue?

**Comment by lezcano** [comment]
This is expected, see https://pytorch.org/docs/master/notes/numerical_accuracy.html
The implementation of numerically stable methods with PyTorch and any eager framework is a task to be solved by the programmer, not the framework.
